course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+#include <stdexcept>
+
+namespace linked_list
+{
+
+// This is a very helpful website: https://www.cs.northwestern.edu/~riesbeck/programming/c++/stl-iterator-define.html#TOC23
+template<class T> class LinkedListIter;
+
+template<class T, class Allocator = std::allocator<T>>
+class LinkedList
+{
+public:
+    friend class LinkedListIter<T>;
+    typedef LinkedListIter<T> iterator;
+
+    // Basic assignment constructors
+    LinkedList() = default;
+    LinkedList(const T& val) // TODO T&& val ?
+    {
+        std::cout << "Basic construction: from " << val << std::endl;
+        data = std::make_unique<T>(val);
+        mSize += 1;
+    }
+    // Copy construction
+    LinkedList(const LinkedList& src)
+    {
+        if (src.data)
+        {
+            std::cout << "Copy construction: from " << src.front() << std::endl;
+            data = std::make_unique<T>(src.front());
+            std::cout << "copy -- set data" << std::endl;
+        }
+        else
+        {
+            std::cout << "Copy construction: from empty" << std::endl;
+        }
+
+        if (src.next)
+        {
+            std::cout << "copy -- setting next" << std::endl;
+            next = std::make_unique<LinkedList<T>>(*src.next.get());
+            std::cout << "copy -- set next" << std::endl;
+        }
+        else
+            std::cout << "copy -- no next" << std::endl;
+        mSize = src.mSize;
+        std::cout << "copy -- set size" << std::endl;
+    }
+    // Move construction
+    LinkedList(LinkedList&& src)
+    {
+        std::cout << "Move construction: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+    }
+    // Copy assignment
+    LinkedList& operator=(const LinkedList& src)
+    {
+        std::cout << "Copy assignment: from " << src.front() << std::endl;
+        data = std::make_unique<T>(src.front());
+        if (src.next)
+            next = std::make_unique<LinkedList<T>>(*src.next.get());
+        mSize = src.mSize;
+        return *this;
+    }
+    // Move assignment
+    LinkedList& operator=(LinkedList&& src)
+    {
+        std::cout << "Move assignment: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+        return *this;
+    }
+
+    bool operator==(LinkedList& cmp) const
+    {
+        std::cout << mSize << " " << cmp.mSize << std::endl;
+        if (mSize != cmp.mSize)
+            return false;
+        if (data != cmp.data)
+            return false;
+        if (!(next && cmp.next))
+             return true;
+        if  (next != cmp.next)
+            return false;
+        return true;
+    }
+
+    int size() const { return mSize; }
+    T front() const // TODO should this be a pointer? const?
+    {
+        if (data)
+            return *data.get();
+        throw std::range_error("Calling front on empty LinkedList is undefined.");
```

Comment:
This is not true. Calling ```front``` on your list produces a behaviour that's very well defined: it throws an exception.
If you'd remove the if/throw, than you can call it 'undefined', and you'd also move the burden of checking if the list is not empty to the user. This would prevent the user for "paying for what they don't use"